okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java

```diff
@@ -376,8 +387,78 @@ private Request createTunnelRequest() {
   /** Returns true if this connection can carry a stream allocation to {@code address}. */
   public boolean isEligible(Address address) {
     return allocations.size() < allocationLimit
-        && address.equals(route().address())
-        && !noNewStreams;
+        && !noNewStreams
+        && canCarryAddress(address);
+  }
+
+  private boolean canCarryAddress(Address address) {
+    boolean directMatch = address.equals(route().address());
+
+    if (directMatch) {
+      return true;
+    }
+
+    if (address.isCoalescable() && this.isCoalescable() && address.equalsNonHost(route.address())) {
+      return supportsHost(address);
+    }
+
+    return false;
+  }
+
+  private boolean isCoalescable() {
+    return http2Connection != null && subjectAlternativeNames != null
+        && route().address().isCoalescable();
+  }
+
+  private boolean supportsHost(Address address) {
+    String addressHost = address.url().host();
+
+    Boolean supportsHost = supportedHostCache.get(addressHost);
+
+    if (supportsHost != null) {
+      return supportsHost;
+    }
+
+    // attempt to coalesce secure HTTP/2 connections by subjectAltNames
+    // currently does not apply to HTTP/1.1 because reuse assumptions are not the same
+    // e.g. app may use domain sharding for performance
+    boolean hostMatch = false;
+    for (String subjectAlternativeName : subjectAlternativeNames) {
+      if (OkHostnameVerifier.verifyHostname(addressHost, subjectAlternativeName)) {
+        hostMatch = true;
+        break;
+      }
+    }
+    if (!hostMatch) {
+      supportedHostCache.put(addressHost, false);
```

Comment:
Yep, I guess caching should come from a benchmark with realistic traffic pattern